processing/src/main/java/io/druid/segment/FilteredOffset.java

```diff
@@ -113,14 +88,32 @@ public boolean withinBounds()
   public void reset()
   {
     baseOffset.reset();
+    incrementIfNeededOnCreationOrReset();
+  }
+
+  private void incrementIfNeededOnCreationOrReset()
+  {
+    if (baseOffset.withinBounds()) {
+      if (!filterMatcher.matches()) {
+        increment();
+        // increment() returns early if it detects the current Thread is interrupted. It will leave this
+        // FilteredOffset in an illegal state, because it may point to an offset that should be filtered. So must to
+        // call BaseQuery.checkInterrupted() and thereby throw a QueryInterruptedException.
+        BaseQuery.checkInterrupted();
+      }
+    }
+  }
+
+  @Override
+  public ReadableOffset getBaseReadableOffset()
+  {
+    return baseOffset.getBaseReadableOffset();
   }
 
   @Override
   public Offset clone()
   {
-    FilteredOffset offset = (FilteredOffset) super.clone();
-    offset.baseOffset = offset.baseOffset.clone();
-    return offset;
+    throw new UnsupportedOperationException("FilteredOffset should not be cloned");
```

Comment:
@leventov why is this changed to throw an exception? More precisely, why are offsets cloned in topN?

Reply:
This code was never called and in fact it was wrong, because `filterMatcher` was not cloned. 

Offset.clone() is used for two purposes:
 1) making "frames" in QueryableIndexStorageAdapter
 2) (in theory) it helps JVM to figure out that the object could be scalarized on stack, within XxxPrototype.scanAndAggregate(). This method is inlined and JVM analyzes the scope of this method. If offset objects comes as parameter, it couldn't be scalarized. If it is cloned in the body of the method, it probably could be.

I don't remember was I actually successful to make JVM to scalarize this offset object or not, through.

Reply:
This causes some queries to fail-- raised an issue https://github.com/druid-io/druid/issues/5132.